csrc/multidevice/executor.cpp

```diff
@@ -6,133 +6,234 @@
  */
 // clang-format on
 #ifdef NVFUSER_DISTRIBUTED
+#include <ATen/cuda/CUDAContext.h>
+#include <device_lower/utils.h>
+#include <fusion_segmenter.h>
 #include <ir/utils.h>
-#include <multidevice/allocator.h>
+#include <multidevice/device_mesh.h>
 #include <multidevice/executor.h>
 #include <multidevice/lower_communication.h>
-#include <multidevice/pipeline.h>
+#include <multidevice/utils.h>
 
 namespace nvfuser {
 
-bool PipelineExecutor::shouldRun(PipelineStage* stage) {
-  if (should_run_.find(stage) == should_run_.end()) {
-    should_run_.emplace(
-        stage, stage->descriptor()->mesh.has(runtime_.comm_.deviceId()));
+namespace {
+
+std::pair<std::unique_ptr<Fusion>, std::unordered_map<Val*, Val*>>
+copyFusionAndChangeOutputs(Fusion* fusion, std::unordered_set<Val*> outputs) {
+  std::unique_ptr<Fusion> fusion_copy = std::make_unique<Fusion>();
+  std::unordered_map<Val*, Val*> copy_to_original_map;
+  auto original_to_copy_cloner = Fusion::copy(fusion, fusion_copy.get());
+
+  auto original_inputs = fusion_copy->inputs();
+  auto original_outputs = fusion_copy->outputs();
+
+  // Remove original outputs
+  std::for_each(
+      original_outputs.begin(), original_outputs.end(), [&](auto& output) {
+        fusion_copy->removeOutput(output);
+      });
+
+  // Add new outputs
+  std::for_each(outputs.begin(), outputs.end(), [&](Val* const& output) {
```

Comment:
I'm not exactly sure what this function is meant to do, but since `outputs` is an unordered map, the order of the outputs will be non-deterministic, which should be avoided.

Reply:
For our case it shouldn't matter that the order is non-deterministic. Also since the output container of both the original and the copy are unordered, I don't see why this is a problem.

This is mimicked on `SegmentedFusion::makeFusion`

I am ok to change according to your suggestion but I don't see how. Let me know.

Reply:
`Fusion::outputs()` returns a deterministically ordered `std::vector`. 

Doesn't the non-determinism here mean generated code can be non-deterministic? That shouldn't be a problem mathematically, but we do not want unnecessary non-determinism.

Reply:
ok, I changed to type of `ouputs` (the argument of `copyFusionAndChangeOutputs`) from `unordered_set` to `set`. Does it address your concern ?

Reply:
No, I don't think so since the order in `std::set` should be just defined by object addresses, so it won't be deterministic.

Reply:
Just use `std::vector`